oslo_messaging/rpc/server.py

```diff
@@ -160,20 +160,34 @@ def _process_incoming(self, incoming):
         except Exception:
             LOG.exception(_LE("Can not acknowledge message. Skip processing"))
             return
-
+        
+        if self.conf.statsd_enabled:
+            self.metrics.queue.put({
+                'message': message,
+                'tag': "event",
+            })
         failure = None
         try:
             res = self.dispatcher.dispatch(message)
         except rpc_dispatcher.ExpectedException as e:
             failure = e.exc_info
             LOG.debug(u'Expected exception during message handling (%s)', e)
+            if self.conf.statsd_enabled:
+                self.metrics.queue.put({
+                    'message': message,
+                    'tag': "expected_exception",
+                })
         except Exception:
             # current sys.exc_info() content can be overridden
             # by another exception raised by a log handler during
             # LOG.exception(). So keep a copy and delete it later.
             failure = sys.exc_info()
-            LOG.exception(_LE('Exception during message handling'))
-
+            LOG.exception('Exception during message handling')
+            if self.conf.statsd_enabled:
+                self.metrics.queue.put({
+                    'message': message,
+                    'tag': "exception",
+                })
```

Comment:
This and the above `put` are blocking calls. I don't think we want to block here. Instead, we either have to use `put_nowait()` or `put(…, block=False)` to not block processing of the message if the queue is full. This could then raise [`Queue.Full`](https://docs.python.org/2/library/queue.html#Queue.Full), which we should handle and log as warning at least.

Reply:
good catch!
```
  try:
      self.metrics.queue.put_nowait({
          'message': message,
          'tag': "event",
      })
  except queue.Full:
      LOG.warning("Statsd metrics queue is full")
```

Reply:
Looks good to me, except that we must use `Queue` for Python 2 it seems:
```
>>> import queue
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named queue
>>> import Queue
>>> Queue.Full
<class 'Queue.Full'>
>>> 
```